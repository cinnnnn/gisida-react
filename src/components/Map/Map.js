import React, { Component } from 'react';
import { connect } from 'react-redux'
import { formatNum, getLastIndex } from '../../utils'
import { processLayer,   } from 'gisida';
import './Map.css';

const mapboxgl = require('mapbox-gl');

const mapStateToProps = (state, ownProps) => {
  return {
    mapConfig: state.APP.mapConfig,
    accessToken: state.APP.accessToken,
    layers: state.LAYERS
  }
}

class Map extends Component {

  constructor(props) {
    super(props);

    // todo - bind all the functions
    // todo - move state to store
    this.state = 
    {
      loaded: false
    }
  }

  initMap(accessToken, mapConfig) {
    if (accessToken && mapConfig) {
      mapboxgl.accessToken = accessToken;
      this.map = new mapboxgl.Map(mapConfig);
      this.map.addControl(new mapboxgl.NavigationControl());
      this.map.on('load', () => {
        this.addMouseEvents();
        // todo - dispatch MAP_LOADED action, triggering default layers to process
      });
    }  
  }

  addMouseEvents() {
    // this.addMapClickEvents()
    // this.addMouseMoveEvents()
    // etc
  }
  
  componentWillReceiveProps(nextProps) {
    // todo - remove this
    if (!this.state.loaded) {
      this.initMap(nextProps.accessToken, nextProps.mapConfig);
    }
    // todo - handle adding and removing layers based on props
  }

  componentDidUpdate(prevProps, prevState) {
    // todo - handle this is in componentWillReceiveProps
   
  }

  componentDidMount() {
    this.initMap(this.props.accessToken, this.props.mapConfig);
  }

  // todo - handle removing layers
  // todo - handle timeseries changes
  // todo - create timeseriesSliderContainer
  // todo - create legendContainer

  // todo - HANDLE THIS.PROPS.CHILDREN

  render() {
    return (
      <div id='map' />
    );
  }
}

export default connect(mapStateToProps)(Map);
