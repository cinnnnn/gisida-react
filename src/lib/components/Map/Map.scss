.marker-label {
  color: #444;
  text-shadow: -1px -1px 0 rgba(255, 255, 255, 0.5), 1px -1px 0 rgba(255, 255, 255, 0.5), -1px 1px 0 rgba(255, 255, 255, 0.5), 1px 1px 0 rgba(255, 255, 255, 0.5);
  font-size: 16px;
  font-weight: bold;
  line-height: 0.9em;
}
