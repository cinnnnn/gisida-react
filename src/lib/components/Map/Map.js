import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Actions, addLayer } from 'gisida';
import './Map.scss';

const mapStateToProps = (state, ownProps) => {
  return {
    mapConfig: state.APP.mapConfig,
    accessToken: state.APP.accessToken,
    layers: state.LAYERS,
    isLoaded: state.MAP.isLoaded,
    styles: state.STYLES,
    regions: state.REGIONS
  }
}

class Map extends Component {

  initMap(accessToken, mapConfig) {
    if (accessToken && mapConfig) {
      mapboxgl.accessToken = accessToken;
      this.map = new mapboxgl.Map(mapConfig);
      this.map.addControl(new mapboxgl.NavigationControl());
      this.map.on('load', () => {
        this.addMouseEvents();
        this.props.dispatch(Actions.mapLoaded(true));
      });
      let theMap = this.map; 

      this.map.on('mousedown', function (e) {
        document.getElementById('info').innerHTML =
          // e.point is the x, y coordinates of the mousemove event relative
          // to the top-left corner of the map
          JSON.stringify(e.point) + '<br />' +
          // e.lngLat is the longitude, latitude geographical position of the event
          JSON.stringify(e.lngLat) + '<br />' +
          'Zoom: ' + theMap.getZoom()  + '<br />' +
          'Center: ' + theMap.getCenter() + '<br />';
      })
    }  
  }

  addMouseEvents() {
    // this.addMapClickEvents()
    // this.addMouseMoveEvents()
    // etc
  }
  
  componentWillReceiveProps(nextProps) {
    const isLoaded = nextProps.isLoaded;
    const layers = nextProps.layers;
    const accessToken = nextProps.accessToken;
    const mapConfig = nextProps.mapConfig;
    const styles = nextProps.styles;
    const regions = nextProps.regions;
  
    // Check if map isLoad and initialize.
    if (!isLoaded) {
      this.initMap(accessToken, mapConfig);
    }

    this.map.on('styledata', () => {
     // Handle Style Changes
    });

    // Check for changes in Map Style
    styles.forEach((style) => {
      if (style.current && mapConfig.style !== style.style) {
        this.map.setStyle(style.style);
      }
    });

    // Add Layers to map from props
    if (isLoaded && layers && Object.keys(layers).length > 0) {
      Object.keys(layers).forEach((key) => {
        const layer = layers[key];
        if (layer.loaded) {
          addLayer(this.map, layer, mapConfig);
        }
      });
    }

    // Set region from props
    regions.forEach((region) => {
      if (region.current) {
        this.map.easeTo({ center: region.center, zoom: region.zoom })
      }
    });
    window.GisidaMap = this.map;
  }

  render() {
    return (
      <div>
        <div id='map' />
        <pre id='info'></pre>
      </div>  
    );
  }
}

export default connect(mapStateToProps)(Map);
