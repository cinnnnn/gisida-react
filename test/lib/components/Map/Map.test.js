import React from 'react';
import { Map }  from '../../../../src/lib/components/Map/Map.js'
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

//not running because mapboxgl is not defined

const style = {width:"100%",display:"inline"};

const componentWrapper = shallow(
	<Map 
		mapId='map-1'
		style={style}
	/>
);

describe('Map', () => {
	it('Map component renders correctly', () => {
		const json = toJson(componentWrapper)
		expect(json).toMatchSnapshot();
	})
});