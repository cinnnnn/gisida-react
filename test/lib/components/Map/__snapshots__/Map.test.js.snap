// Jest Snapshot v1, https://goog.gl/fbAQLP

exports[`Map Map component renderes correctly 1`] = `
  <div>
    {isIE || !mapboxgl.supported() 
      ?
      (
        <div className="alert alert-info">
          Your browser is not supported. Please open link in another browser e.g Chrome or Firefox
        </div>
      ) :
      (
        <div id={this.props.mapId} style={{
          width: mapWidth,
          display: this.props.MAP.blockLoad ? 'none' : 'inline'
        }} >
          <div className="widgets">
            {/* Render Children elemets with mapId prop added to each child  */}
            {
              React.Children.map(this.props.children, child => {
                return React.cloneElement(child, {
                  mapId: this.props.mapId
                });
              })
            }
          </div>
        </div>
      )}
  </div>
`;

//if statements at the beginning just setting values, not JSX stuff
//values are not updated